### DISCUSSION TITLE
Can't get rid of obs-studio-hook

### DISCUSSION THREAD

**Discussion by igor725** [obsproject/obs-studio#10619] [General]
I uninstalled OBS from my computer but the graphics-hook is still there. I tried to manually delete it (the whole `obs-studio-hook` directory) and fix my Vulkan layers configuration nad it looked like everything was normal... until I rebooted my PC. The hook installed itself back again. And this is not a virus or anything, these files actually signed. I also tried to install the latest OBS version and immediately uninstall it to see if issue is fixed... nope. This is annoying a little because my Vulkan apps sometimes hangs at boot and I can't even kill them, because of driver lock (the call stack leads from nvidia driver to OBS graphics hook). Doubt they hang because of it, it looks like transparent pass but I want to remove it anyways. The only solution in this case is to reboot windows entirely.
I'm actually out of ideas at this point how I should handle this :/

![image](https://github.com/obsproject/obs-studio/assets/40758030/7c8222c3-8537-452c-a725-5a640df0c6cf)
![image](https://github.com/obsproject/obs-studio/assets/40758030/8d3a2676-dae4-4b90-b858-9d4caf7eec98)
![image](https://github.com/obsproject/obs-studio/assets/40758030/774214dc-0d8f-450f-9e3b-2cb3528ee4d3)

**Comment by RytoEX** [comment]
These files are copied here on OBS startup if they do not exist because they are _required_ for Vulkan Capture to work correctly. I've just tested our uninstaller and it definitely does remove them. The only way these would reappear is if you ran OBS again or re-ran our installer afterward.

**Reply by igor725** [reply]
Yes, uninstaller removed them. But when I restarted my PC hook registered itself again, without installed OBS.

**Reply by RytoEX** [reply]
There is no mechanism at all in OBS that re-copies these files with a PC restart. Even if there was, this is not possible if OBS is not installed. I've just tested this by uninstalling OBS, verifying that the files are gone, restarting my PC, and verifying that the files are still gone.

**Comment by igor725** [comment]
Well, I literally just recorded a video how it magically appears in my ProgramData directory after I restart my system. And I have no OBS installed atm.

**Comment by igor725** [comment]
Probably some other app installs it, but it's weird behavior for something that is not OBS actually...

**Reply by RytoEX** [reply]
Then some other app is using our graphics hook, which is not encouraged, but would not be the first time. I'd be interested to know _which_ app is doing it, but this is not something caused by our own application.

**Reply by igor725** [reply]
Well, thank you for the answer. I'll dig this thing further.

**Reply by igor725** [reply]
I still don't know what app is doing it, but I created a file named `obs-studio-hook` in `ProgramData` folder. Looks like it won't happen anymore and my vulkan layers list is clean too, so I'm probably okay with that

**Reply by igor725** [reply]
Well, Process Hacker says `System` process writes these files. Now this is weird...

**Reply by RytoEX** [reply]
I'm pretty sure escalation/admin privileges are required to copy to ProgramData, so that doesn't seem too unusual. If you have any other software that does screen recording, streaming, game clipping, replay saving, or anything similar, especially things set to run on system start (or things that have helpers that run on system start, or things that have services), I would look at those.

**Reply by igor725** [reply]
I have only two apps for this task - ShareX (it uses ffmpeg with ddgrab API) and NVIDIA GeForce Experience. There is literally no more recording things on me. Thank you for your time and answers, but I probably just leave `obs-studio-hook` file there, just checked annd this approach has no side effects. Whatever this app is, it trying to install hook only once at startup and doesn't try to do this in infinite loop.

**Comment by ALCR3W** [comment]
In my case that folder is created by Discord. It seems strange to me, but maybe it was created for screen sharing

**Comment by SubhamBurnwal** [comment]
Why does a Visual Studio build process of a project using very very basic Vulkan refer to this obs-studio-hook? I did not link it anywhere. Is it a dependency somewhere?

![image](https://github.com/user-attachments/assets/51cf6719-471e-4965-ba7c-f44abe32a460)


In case anyone was wondering, here is a verified solution to remove (and prevent reappearance) of these files:-

https://stackoverflow.com/a/63761663/8648276

Upvote that for the great help..

**Reply by igor725** [reply]
This hook links to every graphics process that uses Vulkan. But complete disabling of implicit layers could do more harm than good tbh.

**Comment by koloved** [comment]
I understand why OBS hooks Vulkan/graphics applications from a technical point of view, especially for Game Capture.

But the user experience here is really bad.

In my case, OBS Studio had been uninstalled a long time ago, but this directory was still left on the system:

C:\ProgramData\obs-studio-hook\graphics-hook64.dll

Later, when I tried to install OBS again via winget/UniGetUI, the installer failed with:

Files modified by the installer are currently in use by another application.
Exit code: 6

The confusing part is that OBS was not installed and not running. I had no idea what was blocking the installer.

After using Process Explorer, I found that graphics-hook64.dll was loaded into unrelated applications, including my own app Handy. Handy uses WebView2, which renders through DirectX/Vulkan, so OBS hooks it through the graphics/Vulkan layer. Handy itself does not know about this DLL and does not explicitly depend on OBS.

From a normal user’s perspective, this is a terrible failure mode:

1. OBS was already uninstalled.
2. The OBS hook remained in the system.
3. Random graphics/Vulkan/WebView2 applications loaded it.
4. Reinstalling OBS failed.
5. The installer did not clearly tell me which applications were using the hook.
6. The only practical way to diagnose it was to install Process Explorer and manually search for graphics-hook64.dll.

That is not reasonable for ordinary users. A user may have many Vulkan/WebView2/graphics applications running. Are they expected to find and close every hooked process manually before they can reinstall OBS?

I think OBS should handle this better. For example:

- uninstall should remove or disable the leftover hook/layer properly;
- reinstall should provide a clear list of processes currently using graphics-hook64.dll;
- the installer should explain that the leftover OBS hook is still loaded by other graphics applications;
- ideally, OBS should provide a safer recovery path instead of just failing with exit code 6.

Technically, the hook behavior may be understandable. But when the hook survives uninstall and silently blocks reinstall, the experience becomes extremely confusing.
